src/compiler/scanner.ts

```diff
@@ -3184,37 +3196,19 @@ export function createScanner(
                             }
                         }
                         break;
-                    case CharacterCodes.ampersand:
-                        start = pos;
-                        pos++;
-                        if (charCodeChecked(pos) === CharacterCodes.ampersand) {
-                            pos++;
-                            error(Diagnostics.Operators_must_not_be_mixed_within_a_character_class_Wrap_it_in_a_nested_class_instead, pos - 2, 2);
-                            if (charCodeChecked(pos) === CharacterCodes.ampersand) {
-                                error(Diagnostics.Unexpected_0_Did_you_mean_to_escape_it_with_backslash, pos, 1, String.fromCharCode(ch));
-                                pos++;
-                            }
-                        }
-                        else {
-                            error(Diagnostics.Unexpected_0_Did_you_mean_to_escape_it_with_backslash, pos - 1, 1, String.fromCharCode(ch));
-                        }
-                        operand = text.slice(start, pos);
-                        continue;
                 }
-                if (isClassContentExit(charCodeChecked(pos))) {
+                ch = charCodeChecked(pos);
+                if (isClassContentExit(ch)) {
                     break;
                 }
                 start = pos;
-                switch (text.slice(pos, pos + 2)) { // TODO: don't use slice
-                    case "--":
-                    case "&&":
-                        error(Diagnostics.Operators_must_not_be_mixed_within_a_character_class_Wrap_it_in_a_nested_class_instead, pos, 2);
-                        pos += 2;
-                        operand = text.slice(start, pos);
-                        break;
-                    default:
-                        operand = scanClassSetOperand();
-                        break;
+                if (ch === charCodeChecked(pos + 1) && (ch === CharacterCodes.minus || ch === CharacterCodes.ampersand)) {
+                    error(Diagnostics.Operators_must_not_be_mixed_within_a_character_class_Wrap_it_in_a_nested_class_instead, pos, 2);
+                    pos += 2;
+                    operand = text.slice(start, pos);
+                }

```

Comment:
It would be even more intuitive if we were to rescan the class set as intersection/subtraction when we hit any `--` or `&&`, but it is infeasible without a two-pass scan, the first of which must not emit errors. Unfortunately, I need to give up as it would overcomplicate the implementation.